### ISSUE TITLE
postgres: function and procedure parsing/linting

### ISSUE THREAD

**Issue by wwuck** [sqlfluff/sqlfluff#3093]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff currently treats the internals of a postgres function or procedure as a single string literal.  This means that the function and procedure internals are currently not linted by sqlfluff.

### Expected Behaviour

sqlfluff should be able to properly lint the internals of a postgres function or procedure.  There is a PR that implemented similar functionality for T-SQL functions at https://github.com/sqlfluff/sqlfluff/pull/2155.

### Observed Behaviour

See the sqlfluff test files for postgres dialect.  All function and procedure internals are being parsed as a string literal.

https://github.com/sqlfluff/sqlfluff/blob/4bc2916bf87b260744ba77d2d6a6f5ed407a2ba8/test/fixtures/dialects/postgres/postgres_create_dollar_quoted_function.sql
https://github.com/sqlfluff/sqlfluff/blob/4bc2916bf87b260744ba77d2d6a6f5ed407a2ba8/test/fixtures/dialects/postgres/postgres_create_dollar_quoted_function.yml
https://github.com/sqlfluff/sqlfluff/blob/4bc2916bf87b260744ba77d2d6a6f5ed407a2ba8/test/fixtures/dialects/postgres/postgres_create_function.sql
https://github.com/sqlfluff/sqlfluff/blob/4bc2916bf87b260744ba77d2d6a6f5ed407a2ba8/test/fixtures/dialects/postgres/postgres_create_function.yml
https://github.com/sqlfluff/sqlfluff/blob/4bc2916bf87b260744ba77d2d6a6f5ed407a2ba8/test/fixtures/dialects/postgres/postgres_create_procedure.sql
https://github.com/sqlfluff/sqlfluff/blob/4bc2916bf87b260744ba77d2d6a6f5ed407a2ba8/test/fixtures/dialects/postgres/postgres_create_procedure.yml

### How to reproduce

Parse and lint the sql from the test files for postgres dialect listed under `Observed Behaviour`.

### Dialect

postgres

### Version

sqlfluff, version 0.12.0
Python 3.10.4

### Configuration

```ini
[sqlfluff]
dialect = postgres
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
That would only work for the SQL functions (and not PLPGSQL for example) but even than could be a fair bit of effort because the Postgres Grammar is not complete (and maybe never will be!).

For example, the first file has the following function definition:

```sql
select $1 + $2;
```

This does not parse as a standalone statement as it depends on the two parameters passed in.

Once we stop parsing this as string literal, we need to parse all possible SQL or people's `CREATE FUNCTIONS` will fail. While a string literal is obviously limited, in that it's not checking that function body syntax, it does have the advantage that the surrounding SQL can be parse, linted and fixed without needing to understand that.

Similarly, for second one:
```sql
    BEGIN
        RETURN i + 1;
    END;
```

This is part of PLPGSQL which is not supported by SQLFluff. A few TSQL users contributed a lot of TSQL's control statements (`BEGIN`, `END`, `RETURN`...etc.) but those either have only basic, or no support at all, for Postgres. Again there could be quite an effort to bring those up to the necessary speed to enable PLPGSQL support.

I'm not saying this is impossible, or will never be done, but would need a lot of effort and contributions from the community IMHO.

**Comment by barrywhart** [comment]
In line with the discussion above, changed this issue from bug to enhancement.

**Comment by simonjobs** [comment]
Hi,

I am finding this issue about a year later and I am in a very similar situation although we are running Redshift.

So I am wondering if there has been any progress in linting stored procedures and PLPGSQL specifically or if the situation still looks the same? @wwuck what path did you end up taking?

**Comment by WittierDinosaur** [comment]
@simonjobs I think this is still unlikely to be picked up any time soon. Happy to review if someone raises a PR, but this is a non-trivial change, and I don't think any of the maintainers plan on working on it

**Comment by enote-kane** [comment]
As much as I would love to see this feature as well, especially with PostgreSQL this becomes a whole new feature set.

I have functions written in at least SQL, PL/pgSQL, Python and Rust and only SQL is an actual dialect which SQLFluff could possibly deal with.

For the others, we would need an actual child execution with completely different linters (external programs/Python modules, not just dialects) and it also comes with the issues of the context (input/output variables, data type mapping, ...), linter configuration, aso.

**Comment by manmartgarc** [comment]
Hey folks, is there any work started on this? I'd love to contribute to this.

**Comment by daniel-salgado** [comment]
I’m very interested on this issue. Working a lot with PostegreSQL lately and would be awesome using the SQLFluff  for it.

**Comment by peterbud** [comment]
Postgres anonymous code block can be in any language according to the documentation, not just SQL or plpqsql: https://www.postgresql.org/docs/current/sql-do.html

Sqlfluff won't parse these blocks and treat them as literal strings, and therefore won't lint them. There is no roadmap to provide external linter hand-off and child execution to python or similar linters.
